### DISCUSSION TITLE
Branch protections when actions use `paths-ignore`

### DISCUSSION THREAD

**Discussion by kevinm416** [community/community#54877] [Actions] [answered]
### Select Topic Area

Question

### Body

I've got a repo with a frontend build and a server build. Each only runs when files in the appropriate folder change via the `paths-ignore` option in the github workflow files. I'm trying to configure my repo so that merges to master can only happen if whatever checks are triggered have passed. 

I don't see an option for doing that though. It looks like I can only set up the branch rule so that a hard-coded list of checks pass. If I put all the checks in the list that need to pass, then a web-only change would never be able to be merged because the server build never runs. 

This seems like it would be a common situation. Has anyone figured out how to do this?

**Comment by hrishikesh-kocharekar-dev** [comment]
Hi Kevin, 
Yes, it's a common situation to have multiple checks that need to pass before merging to the master branch. Fortunately, GitHub provides a flexible way to configure branch protection rules to enforce certain requirements.

To achieve what you are describing, you can set up a branch protection rule for the master branch that requires all checks to pass before allowing a merge. Then, you can add multiple status checks for your repository in the settings. These status checks can be configured to only run on certain files or folders by using the "paths-ignore" option in the GitHub workflow files.

Here are the general steps to set up the branch protection rule and status checks:

Go to your repository on GitHub and click on "Settings".
Click on "Branches" on the left-hand side menu.
Under "Branch protection rules", select the master branch and click "Edit".
Check the box next to "Require status checks to pass before merging".
Select the status checks you want to require from the list.
Click "Save changes".
Now, when someone tries to merge changes into the master branch, all the selected status checks will be run. If any of the checks fail, the merge will be blocked until the issues are resolved.

To make sure that your status checks only run on the appropriate files, you can configure the GitHub workflow files in your repository to use the "paths-ignore" option. For example, you could have one workflow file that only runs when changes are made to the frontend files and another workflow file that only runs when changes are made to the server files.

I hope this helps!

**Reply by kevinm416** [reply]
Hi Hrishikesh, thank you for your reply, However, I already have all the stuff you described set up. Maybe this is a clearer explanation of my issue.

When I configure the checks like this, everyone can merge without the checks completing successfully:

![Screenshot from 2023-05-08 11-21-20](https://user-images.githubusercontent.com/526698/236901255-e4822b4a-bfcf-4300-a93e-6f47633d0623.png)

When I configure the checks like this, I can't merge frontend changes, because they do not trigger `build-api`:

![Screenshot from 2023-05-08 11-21-31](https://user-images.githubusercontent.com/526698/236901392-1f02e073-414c-470d-afd1-dd79b518f0cb.png)

**Reply by danielo515** [reply]
How do you exclude certain PRs from requiring certain checks?

**Comment by fniephaus** [comment, marked as answer]
This is probably what you are looking for: https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/managing-protected-branches/troubleshooting-required-status-checks#handling-skipped-but-required-checks

**Reply by kevinm416** [reply]
Hi Fabio, thanks for the pointer! I actually read that at the time, but didn't really understand what it was suggesting. 

What I've done is require a `pre-merge` check before merging to the main branch, and added that `pre-merge` check in each build workflow. Previously, the checks were named things like `build-api`, `build-web`, etc. By naming them all `pre-merge`, whichever workflows get triggered, at least one will provide a `pre-merge` result, and it's ok if multiple provide `pre-merge` checks.

I think that's what that article is hinting at, but it didn't click for me.

**Reply by ecerulm** [reply]
The link above is broken as of 2026-07-22 

I assume the new location is https://docs.github.com/en/pull-requests/how-tos/merge-and-close-pull-requests/troubleshooting-required-status-checks#handling-skipped-but-required-checks

<img width="860" height="452" alt="image" src="https://github.com/user-attachments/assets/e4107e4b-6c0e-4da6-b86b-388b3a46d60e" />

**Comment by rob-levy-minimum** [comment]
The link in the accepted answer is a 404, but here's [an updated link](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/collaborating-on-repositories-with-code-quality-features/troubleshooting-required-status-checks#handling-skipped-but-required-checks).

**Reply by ViacheslavP** [reply]
FYI The previous link contained the workaround that doesn't seem to work today: http://web.archive.org/web/20230506221415/https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#onpushpull_requestpull_request_targetpathspaths-ignore

**Reply by bojingo** [reply]
The content at the link was updated again but only gives an explanation and example of why the PRs get stuck "pending" on an action that was not run due to a path filter, but doesn't fully explain or give an example of a solution! I can infer to use a conditional to skip execution rather than path filter, but how about giving us an example and explanation of a working solution to the problem?

Azure Repos gives ability to set path filter in build validations (roughly equivalent of GitHub Actions workflow branch protection status check). I'd like to know how to accomplish the same capability.

**Reply by PabloReszczynski** [reply]
I'm on the same boat. I'm trying to see if any of the [context-provided values](https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/accessing-contextual-information-about-workflow-runs#context-availability) give any information on what files changed but I can't find any.

**Comment by wood-jp** [comment]
From the [linked docs](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/collaborating-on-repositories-with-code-quality-features/troubleshooting-required-status-checks#handling-skipped-but-required-checks):


> Warning
> 
> If a workflow is skipped due to [path filtering](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#onpushpull_requestpull_request_targetpathspaths-ignore), [branch filtering](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#onpull_requestpull_request_targetbranchesbranches-ignore) or a [commit message](https://docs.github.com/en/actions/managing-workflow-runs/skipping-workflow-runs), then checks associated with that workflow will remain in a "Pending" state. A pull request that requires those checks to be successful will be blocked from merging.
> 
> If, however, a job within a workflow is skipped due to a conditional, it will report its status as "Success". For more information, see [Using conditions to control job execution](https://docs.github.com/en/actions/using-jobs/using-conditions-to-control-job-execution).

However, this in no way helps to explain what the correct solution is for the very example provided below it:

```
name: ci
on:
  pull_request:
    paths:
      - 'scripts/**'
```

Supposing that there exists some build/test job here, and that job should be run as shown in the example when any change happens other than something in the `scripts/` directory. Moreover, it should be a **required check (when it is run)**. The problem remains that setting this check to be required would cause a change to only `scripts/` to be forever rejected. How would one use conditionals to solve this exact situation? Reading those docs I have come away with no clear idea.

Ideally there would be a way to indicate (per job) if a skipped job should just return success or failure rather than remain forever pending.
